htlcswitch/quiescer.go

```diff
@@ -0,0 +1,160 @@
+	remoteInit bool
+
+	// sent tracks whether or not we have emitted Stfu for sending.
+	sent bool
+
+	// received tracks whether or not we have received Stfu from our peer.
+	received bool
+}
+
+// newQuiescer creates a new quiescer for the given channel.
+func newQuiescer(cfg quiescerCfg) quiescer {
+	return quiescer{
+		quiescerCfg: cfg,
+	}
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescer) recvStfu(msg lnwire.Stfu) error {
+	if q.received {
+		return fmt.Errorf("%w for channel %v", ErrStfuAlreadyRcvd,
+			q.chanID)
+	}
+
+	q.received = true
+
+	// If the remote party set the initiator bit to true then we will
+	// remember that they are making a claim to the initiator role. This
+	// does not necessarily mean they will get it, though.
+	q.remoteInit = msg.Initiator
+
+	return nil
+}
+
+// sendStfu is called when we are ready to send an Stfu message. It returns the
+// Stfu message to be sent.
+func (q *quiescer) sendStfu() fn.Result[lnwire.Stfu] {
+	if q.sent {
+		return fn.Err[lnwire.Stfu](
+			fmt.Errorf("%w for channel %v", ErrStfuAlreadySent,
+				q.chanID),
+		)
+	}
+
+	stfu := lnwire.Stfu{
+		ChanID:    q.chanID,
+		Initiator: q.localInit,
+	}
+
+	q.sent = true
+
+	return fn.Ok(stfu)
+}
+
+// oweStfu returns true if we owe the other party an Stfu. We owe the remote an
+// Stfu when we have received but not yet sent an Stfu.
+func (q *quiescer) oweStfu() bool {
+	return q.received && !q.sent
+}
+
+// needStfu returns true if the remote owes us an Stfu. They owe us an Stfu when
+// we have sent but not yet received an Stfu.
+func (q *quiescer) needStfu() bool {
+	return q.sent && !q.received
+}
+
+// isQuiescent returns true if the state machine has been driven all the way to
+// completion. If this returns true, processes that depend on channel quiescence
+// may proceed.
+func (q *quiescer) isQuiescent() bool {
+	return q.sent && q.received
+}
+
+// downstreamLeader determines which ChannelParty is the initiator of quiescence
+// for the purposes of downstream protocols. If the channel is not currently
+// quiescent, this method will return None.
+func (q *quiescer) downstreamLeader() fn.Result[lntypes.ChannelParty] {
+	if !q.isQuiescent() {
+		return fn.Err[lntypes.ChannelParty](ErrNoDownstreamLeader)
+	}
+
+	// We assume it is impossible for both to be false, if the channel is
+	// quiescent. However, we use the same tie-breaking scheme no matter
+	// what.
+	if q.localInit == q.remoteInit {
+		return fn.Ok(q.channelInitiator)
+	}
+
+	// In this case we know that only one of the values is set so we just
+	// return the one that indicates whether we end up as the initiator.
+	if q.localInit {
+		return fn.Ok(lntypes.Local)
+	}
+
+	return fn.Ok(lntypes.Remote)
+}
+
+// canSendUpdates returns true if we haven't yet sent an Stfu which would mark
+// the end of our ability to send updates.
+func (q *quiescer) canSendUpdates() bool {
+	return !q.sent
```

Comment:
We later on add `q.localInit` to the state machine when we add the ability to kick off a quiescence session - I think we can add it in this commit to complete the state machine.